### PR TITLE
[RFC] Prometheus query scraper for etcd metrics

### PR DESCRIPTION
**Pull request by hawkowl**
### Which issue this PR addresses:

Part of https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/10729300/

### What this PR does / why we need it:

Runs Prometheus queries for use in the monitor component

### Test plan for issue:

Probably some unit tests when it's more solidified, but I ran it locally against v4-shared as a test.

### Is there any documentation that needs to be updated for this PR?

Probably if we decide to do it

### CODE REVIEW

**Comment by nilsanderselde** [comment]
We considered doing this in https://github.com/Azure/ARO-RP/pull/1430.

There were several concerns. See Jim's comment in [9534770](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9534770).

>The equivalent integration in v3 was a mess:
>- huge amount of manual effort in building and maintaining a list of select metrics with pruned dimensions
>- even more effort defining Geneva preaggregates.  In v3 we never really even got as far as meaningful dashboards.
>- poor fit between prometheus and Geneva hot-path models, so complexity in managing appropriate translations
>- we overloaded our Geneva quota, and v3 was orders of magnitude smaller than v4 is now
>
>In the v4 world, add that I dislike the idea of pulling metrics via the monitor - it's not scalable.

Are these not problems anymore?

**Comment by hawkowl** [comment]
> We considered doing this in #1430.
> 
> There were several concerns. See Jim's comment in [9534770](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9534770).
> 
> > The equivalent integration in v3 was a mess:
> > 
> > * huge amount of manual effort in building and maintaining a list of select metrics with pruned dimensions
> > * even more effort defining Geneva preaggregates.  In v3 we never really even got as far as meaningful dashboards.
> > * poor fit between prometheus and Geneva hot-path models, so complexity in managing appropriate translations
> > * we overloaded our Geneva quota, and v3 was orders of magnitude smaller than v4 is now
> > 
> > In the v4 world, add that I dislike the idea of pulling metrics via the monitor - it's not scalable.
> 
> Are these not problems anymore?

I think that etcd problems are causing enough problems that some metrics, even if collected infrequently, will give us some warning and capability before issues occur. Those are definitely problems, but until we design the egress lockdown metrics thing-y, we don't particularly have much option...

File: pkg/monitor/cluster/etcdmetrics.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	"github.com/Azure/ARO-RP/pkg/proxy/prometheus"
+	"github.com/Azure/ARO-RP/pkg/util/roundtripper"
+)
+
+func (mon *Monitor) emitEtcdMetrics(ctx context.Context) error {
+	var etcdObjectCount int64 = -1
+
+	prom := prometheus.NewPrometheusProxyWithRestConfig(mon.log, mon.restconfig)
+
+	promClient, err := api.NewClient(api.Config{Address: "http://prometheus-k8s-0:9090", RoundTripper: roundtripper.RoundTripperFunc(prom.RoundTripper)})
```
Comment (gvanderpotte) [inline pkg/monitor/cluster/etcdmetrics.go:23]: We could have a retry with k8s-1 too

File: pkg/monitor/cluster/cluster.go [hunk 2]
```diff
@@ -135,6 +135,7 @@ func (mon *Monitor) Monitor(ctx context.Context) (errs []error) {
 		mon.emitStatefulsetStatuses,
 		mon.emitJobConditions,
 		mon.emitSummary,
+		mon.emitEtcdMetrics,
```
Comment (gvanderpotte) [inline pkg/monitor/cluster/cluster.go:138]: In addition to this, we could also push metrics about node cpu utilization and memory utilization. If we start pulling metrics from prometheus and push them to geneva, then we might as well take those.

**Review by gvanderpotte — COMMENTED** [review]
Overall, I think having more key data is a good idea which would give us more visibility and be able to anticipate problematic. While thinking of it, my concern is that we will have a dependency on prometheus, which is on running on worker nodes (and might get disrupted by interactions with what's running there).

**Pull request closed without merging**